nupic/research/frameworks/vernon/mixins/gradient_metrics.py

```diff
@@ -0,0 +1,232 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2021, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import abc
+from copy import deepcopy
+from pprint import pformat
+
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+
+from nupic.research.frameworks.pytorch.hooks import ModelHookManager, TrackGradientsHook
+from nupic.research.frameworks.pytorch.model_utils import filter_modules
+
+
+class GradientMetrics(metaclass=abc.ABCMeta):
+    """
+    Mixin for tracking and plotting module gradient metrics during training.
+
+
+    :param config: a dict containing the following
+
+        - gradient_metrics_args: a dict containing the following
+
+            - include_modules: (optional) a list of module types to track
+            - include_names: (optional) a list of module names to track e.g.
+                             "features.stem"
+            - include_patterns: (optional) a list of regex patterns to compare to the
+                                names; for instance, all feature parameters in ResNet
+                                can be included through "features.*"
+            - plot_freq: (optional) how often to create the plot, measured in training
+                         iterations; defaults to 1
+            - metrics: a list of metrics options from ["cosine", "dot",
+            "pearson"]; defaults to ["cosine",]
+            - gradient_transformation: (optional) can be one of "real", "sign",
+            "mask". "real" corresponds to the real values of the gradients,
```

Comment:
Please fix hanging indents for 'metrics' and 'gradient_transformation'. Do the same for 'max_samples_to_track' (below).

Reply:
In fact, you started the new line too early.